Fix misleading docs around `--reload` behaviors (#1331)
Kludex/uvicorn: docs/settings.md

## Development

> * `--reload` - Enable auto-reload. Uvicorn supports two versions of auto-reloading behavior enabled by this option. See below for important differences between them.

Reviewer: I can't remember if it was this project that had auto creation of CLI flags using docs, so I'd be weary with the "See below" and replace the last sentence with an explicit explanation. But Kludex would know.

Author: I think I can eliminate the relative reference anyway. Let me try….

Author: Resolved by subsequent commit.

* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default the whole current directory will be watched. If you are running programmatically use `reload_dirs=[]` and pass a list of strings.

> ### Naive Reloading

Reviewer 2: ```suggestion
### Reloading without watchgod
```

Author: Adopted in lastest commit.

> If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to any `*.py` file inside of its monitored directories. See the `--reload-dir` option. Note: _all_ `*.py` files, and _only_ `*.py` files will be monitored for changes. This is not configurable unless watchgod in installed.

Author: @Andrew-Chen-Wang, this adopts your suggested edit and removes what I perceive as extraneous words.

Reviewer: ```suggestion
If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to any `*.py` file inside of its monitored directories (see `--reload-dir`). Note: _all_ `*.py` files, and _only_ `*.py` files will be monitored for changes. This is not configurable unless watchgod in installed.
```

LGTM.

One concern is this quote: "Note: _all_ `*.py` files, and _only_ `*.py`". I thought only `.py` files which were imported would be subject to a reload. This wording implies any `.py` file will be reloaded, regardless of its location in the entire file system.

Author: I might be confused, but unless I'm misreading [the code](https://github.com/encode/uvicorn/blob/6885e9c6ec4d881ec5cf1464a07c619706ad8240/uvicorn/supervisors/statreload.py#L49-L52), I'm pretty sure it's _any_ `*.py` file inside the monitored directories that has a change to its `mtime`. In any event, I word-smithed this a little bit to make clearer that it was bound by the monitored directories.

Either way, we should definitely surface a clearer understanding of what actually happens, so I'm very grateful for the (continued) attention.

> If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to any `*.py` file inside of its monitored directories. See the `--reload-dir` option above. Note: _all_ `*.py` files, and _only_ `*.py` files will be monitored for changes. This is not configurable under naive reloading.

Reviewer: I think it's better to be explicit by mentioning in the last sentence with " unless watchgod is utilized"

```suggestion
If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to any `*.py` file inside of its monitored directories. See the `--reload-dir` option above. Note: _all_ `*.py` files, and _only_ `*.py` files will be monitored for changes. This is not configurable under naive reloading unless watchgod is utilized.
```

Author: Good call. I think I can eliminate even more words in adopting yours.

Author: I'm going to resolve this for now, since I believe subsequent commits resolve this, but please revive if you think I've missed something.

> If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to all `*.py` files (and only `*.py` files) inside of its monitored directories. See the `--reload-dir` option. Which files are monitored is not configurable unless watchgod in installed. See the `--reload-include` and `--reload-exclude` options for details.

Author: @Andrew-Chen-Wang, here's my proposal which is hopefully responsive to [your recent observations](https://github.com/encode/uvicorn/pull/1331#discussion_r786256431) about clarity regarding which `*.py` files would trigger reloads in this context.

Reviewer: Just misspelling (is vs. in installed) and clarity. Otherwise, LGTM!:

```suggestion
If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to all `*.py` files (and only `*.py` files) inside of its monitored directories. See the `--reload-dir` option. Specifying other file extensions is not supported unless watchgod is installed. See the `--reload-include` and `--reload-exclude` options for details.
```

Author: Awesome catch! Thanks!

> ### watchgod-Enabled Reloading

Reviewer 2: ```suggestion
### Reloading with watchgod
```

Author: Adopted in latest commit.

> For more nuanced control over which file modifications trigger reloads, install [watchgod](https://pypi.org/project/watchgod/) where Unvicorn can see it. Alternatively, install `uvicorn[standard]`, which includes watchgod as a dependency. This will enable the following options (which are otherwise ignored).

Reviewer 2: I'd suggest the opposite, 1st install `uvicorn[standard]` and then watchgod manually

Author: I swapped the phrases in the latest commit. (See https://github.com/encode/uvicorn/pull/1331/files#r803758196. Please let me know there if I missed the mark.)

* `--reload-include <glob-pattern>` - Specify a glob pattern to match files or directories which will be watched. May be used multiple times. By default the following patterns are included: `*.py`. These defaults can be overwritten by including them in `--reload-exclude`.
* `--reload-exclude <glob-pattern>` - Specify a glob pattern to match files or directories which will excluded from watching. May be used multiple times. By default the following patterns are excluded: `.*, .py[cod], .sw.*, ~*`. These defaults can be overwritten by including them in `--reload-include`.

> _Note: As of this writing, watchgod relies on poll-based OS machinery to surface modifications to Uvicorn. This means that monitored directories must live on filesystems that support that operation. That_ should _accommodate most use cases. Be aware that for locations where such mechanisms are unavailable, modifications will be silently overlooked._

Reviewer 2: what locations are you thinking of when saying that ?

Author: I don't have a handle on all the weird things that masquerade as filesystems where `poll` or `select` don't maintain appropriate semantics. Whatever those are, they're likely very esoteric setups. I can remove the note, if you think it's not helpful.

Reviewer 2: yes let's get rid of it, no need to go into too much details here we just want users to know how we handle things and this PR clarifies it  quite nicely without that note imho

Author: Removed by latest commit.

## Development

* `--reload` - Enable auto-reload. Uvicorn supports two versions of auto-reloading behavior enabled by this option. There are important differences between them.
* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default the whole current directory will be watched. If you are running programmatically use `reload_dirs=[]` and pass a list of strings.

### Reloading without watchgod

If Uvicorn _cannot_ load [watchgod](https://pypi.org/project/watchgod/) at runtime, it will periodically look for changes in modification times to all `*.py` files (and only `*.py` files) inside of its monitored directories. See the `--reload-dir` option. Specifying other file extensions is not supported unless watchgod is installed. See the `--reload-include` and `--reload-exclude` options for details.

### Reloading with watchgod

> For more nuanced control over which file modifications trigger reloads, install `uvicorn[standard]`, which includes watchgod as a dependency. Alternatively, install [watchgod](https://pypi.org/project/watchgod/) where Unvicorn can see it. This will enable the following options (which are otherwise ignored).

Author: @euri10, does this capture [what you meant](https://github.com/encode/uvicorn/pull/1331#discussion_r803498434)?

Reviewer 2: yep
